sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala

```diff
@@ -282,10 +286,16 @@ class ArtifactManager(session: SparkSession) extends Logging {
     }
   }
 
+  def classloader: ClassLoader = cachedClassLoader.getOrElse {
+    val loader = buildClassLoader
+    cachedClassLoader = Some(loader)
+    loader
+  }
```

Comment:
Synchronized block here to protect against race conditions

Reply:
Done!
I am thinking if `AtomicReference` would be a better choice here, but I see it's missing a `compareAndExchange` variant that updates the current value when it is not equal to the witness value.